### ISSUE TITLE
Synchronisation between pip and conda packages regarding dependencies

### ISSUE THREAD

**Issue by anjos** [conda-forge/torchvision-feedstock#87]
### Solution to issue cannot be found in the documentation.

- [X] I checked the documentation.

### Issue

Since a few generations, we have been noticing that environments created with this conda package require a dependence match that is different than that of the original (PyPI) package.  I wonder if this is a required feature or a bug that could be fixed.

Here is a practical test with a pure Python installation:

```sh
$ python3 -m venv venv
$ source venv/bin/activate
$ pip install torch==2.1.0 torchvision==0.16.1
...
ERROR: Cannot install torch==2.1.0 and torchvision==0.16.1 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested torch==2.1.0
    torchvision 0.16.1 depends on torch==2.1.1
```

However, as this package is made available on the conda-forge channel, this restriction is somehow not respected:

```sh
$ mamba create -n test python=3 pytorch=2.1.0 torchvision=0.16.1
# ...
# installation works! (at least on my macOS test system)
```

Any ideas?

### Installed packages

```shell
N/A
```

### Environment info

```shell
N/A
```

**Comment by hmaarrfk** [comment]
This have been found to be quite compatible between patch versions.

This kinda comes down to the philosophy of the developers.

Upstream pytorch is likely more insterested in strict pinnings that they have rigorously tested and QC'ed.

We generally like to pin looser to allow more compatibility between packages.

Keeping up the pinnings is somewhat a manual process and can lead to errors.

I think now torchvision 0.16.1 should be compatible with pytorch 2.1.2 but it isn't. I'm trying to address this in
https://github.com/conda-forge/torchvision-feedstock/pull/89

**Comment by anjos** [comment]
I understand your concern with a larger usage scope.  I should also thank you for the exceptional work maintaining this feedstock.  I'll check #80.

Regarding compatibility, I'm curious - are reviewers and developers in this feedstock fairly integrated with the pytorch community to understand implications of overriding pins?  If so, shouldn't we just discuss this with upstream and try to change it?  I don't think downstream packaging should be changing pins otherwise - it is quite risky.

There is another issue which relates to reproducibility - you cannot match the version pinning using pip. As of now, we have developers using different install techniques (conda vs. pip), leading to different environments.  Checking issues is already pretty hard when everything else matches, but when there are software stack changes, the surface to be explored increases in size.

With these arguments in mind, could you please reconsider this policy?

**Comment by hmaarrfk** [comment]
Upstream PyTorch developers have been rather silent when I spoke to them. 

> otherwise - it is quite risky.

can you point to real issues? We try to address them when they come up. I also read release notes from time to time. Their patch releases are really minor. 

Having “reproducible” environments across pypi and conda-forge seems really out of scope. It almost sounds like an oxymoron 


Open source is about choice. Feel free to completely ignore conda-forge and follow PyTorch’s standard installation instructions. But clearly there are some advantages that bring you here. 

The truth is as much as I love conda forge. It is unrealistic to expect upstream PyTorch developers to “care” about this project. There are many open source package managers. Each have different goals. PyTorch has their own. They have to focus on what brings them value.

**Comment by anjos** [comment]
> Upstream PyTorch developers have been rather silent when I spoke to them.

It's a pity, indeed...
 
> can you point to real issues? We try to address them when they come up. I also read release notes from time to time. Their patch releases are really minor.

Not immediately.  I'm just being cautions ATM - and annoyed because of the mismatch between environments from our researchers.

That said, I checked the recipe for this package, and I understand you run the complete test suite to verify builds are sensible - kudos on that, as many python packages in conda-forge aren't properly tested.  In this regard, I agree it should be OK to change "patch" pins as suggested.

> Having “reproducible” environments across pypi and conda-forge seems really out of scope. It almost sounds like an oxymoron

> Open source is about choice. Feel free to completely ignore conda-forge and follow PyTorch’s standard installation instructions. But clearly there are some advantages that bring you here.

So far, it has been about getting things done with a wider software set than that accessible by pip, cross-compatibility between packages, readily available CI infrastructure, automation, and open community.  That said, I'll accept whatever choice the various communities can provide, if I cannot manage myself the huge amount of work they're already doing (e.g.: #89).

> The truth is as much as I love conda forge. It is unrealistic to expect upstream PyTorch developers to “care” about this project. There are many open source package managers. Each have different goals. PyTorch has their own. They have to focus on what brings them value.

Very well, it is a fair point.

To sum up, I think you're doing an incredible job already, and I now better understand why there are mismatches. I'll close the issue.  Thanks for all clarifications and the discussion.

**Comment by hmaarrfk** [comment]
beyond upstream version numbers, there are a few things that are different from conda and pip.

The most important one is that pip typically vendors C libraries, whiles conda(-forge) tries to share them.

We are also quite aggressive in versions. For example, I don't think many people have built with ffmpeg 6 whereas we started building with it very shortly after it was released. 

Get your team used to sharing the output of:

```
mamba info
pip list
mamba list
```
when they think of issues and you should be able to resolve issues quickly.

We also started to include runtime checks for versions in our code. When the version mismathes, we include the command to update to the minimum required version.

We only use conda, so writing that command is a little easier.
